File: luiz/ba2a.py
#! /usr/bin/env python

from ba1g import hamming
from ba1i import neighbors


def motif_enumeration(text, k, distance):
    patterns = set()

    first = text[0]
    for i in range(len(first) - k + 1):
        pattern = first[i : i + k]
        for pattern_p in neighbors(pattern, distance):
            patterns.add(pattern_p)

    for dna in text[1:]:
        current_patterns = set()

        for i in range(len(dna) - k + 1):
            for pattern_p in patterns:
                if hamming(dna[i : i + k], pattern_p) <= distance:
                    current_patterns.add(pattern_p)

        patterns = current_patterns

    return patterns


assert motif_enumeration("ATTTGGC TGCCTTA CGGTATC GAAAATT".split(), 3, 1) == set(
    "ATA ATT GTT TTT".split()
)


if __name__ == "__main__":
    with open("data/rosalind_ba2a.txt", "r") as dataset:
        k, d = [int(i) for i in dataset.readline().rstrip().split()]
        text = [s.rstrip() for s in dataset.readlines()]
    print(*motif_enumeration(text, k, d))

File: luiz/src/bin/ba2a.rs
use std::collections::HashSet;
use std::env;
use std::error::Error;
use std::fs;
use std::str;

use bioinformatics_algorithms::{hamming, neighbors};

pub fn motif_enumeration(text: Vec<String>, k: usize, d: usize) -> HashSet<String> {
    let mut patterns = HashSet::new();

    let first = &text[0];
    first
        .as_str()
        .as_bytes()
        .windows(k)
        .map(|pattern| {
            for pattern_p in neighbors(str::from_utf8(pattern).unwrap(), d) {
                patterns.insert(pattern_p);
            }
        })
        .count();

    for dna in &text[1..] {
        let mut current = HashSet::<String>::new();
        dna.as_str()
            .as_bytes()
            .windows(k)
            .map(|x| {
                for pattern_p in &patterns {
                    let pattern = str::from_utf8(x).unwrap();
                    if hamming(pattern, pattern_p) <= d {
                        current.insert(pattern_p.clone());
                    }
                }
            })
            .count();

        patterns = current.into_iter().collect();
    }

    patterns
}

fn main() -> Result<(), Box<dyn Error>> {
    let input: String = env::args()
        .nth(1)
        .unwrap_or("data/rosalind_ba2a.txt".into());
    let data = fs::read_to_string(input)?;
    let mut lines = data.lines();

    let mut ints = lines
        .next()
        .unwrap()
        .split(' ')
        .map(|x| x.parse().unwrap())
        .take(2);
    let k = ints.next().unwrap();
    let d = ints.next().unwrap();

    let text: Vec<String> = lines.map(|x| x.into()).collect();

    for word in motif_enumeration(text, k, d) {
        print!("{} ", word);
    }
    println!();

    Ok(())
}
